url2a908001cd_malictus_klang_tgz-pJ8-malictus_klang_test_PrimitiveTestJ8/src/malictus/klang/chunk/CHRMChunk.java

```diff
@@ -0,0 +1,163 @@
+/**
+ * This file is distributed under a BSD-style license. See the included LICENSE.txt file
+ * for more information. 
+ * Copyright (c) 2009, James Halliday
+ * All rights reserved.
+ */
+package malictus.klang.chunk;
+
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import malictus.klang.file.*;
+import malictus.klang.*;
+import malictus.klang.primitives.*;
+
+/**
+ * A class that represents the CHRM chunk in a PNG file.
+ * @author Jim Halliday
+ */
+public class CHRMChunk extends Chunk {
+	
+	public CHRMChunk(KlangFile parentFile, long startpos, long endpos, long datastartpos, long dataendpos, int chunkType, String chunkName, ContainerChunk parentChunk, RandomAccessFile raf) throws IOException {
+		super(parentFile, startpos, endpos, datastartpos, dataendpos, chunkType, chunkName, parentChunk, raf);
+		raf.seek(datastartpos);
+		/*
+		 * LOAD PRIMITIVE OBJECTS (UNINITIALIZED)
+		 */		
+		Primitive whitepointx = new Int4ByteUnsignedBE();		
+		PrimitiveData whitepointxData;
+		
+		Primitive whitepointy = new Int4ByteUnsignedBE();		
+		PrimitiveData whitepointyData;
+		
+		Primitive redx = new Int4ByteUnsignedBE();		
+		PrimitiveData redxData;
+		
+		Primitive redy = new Int4ByteUnsignedBE();		
+		PrimitiveData redyData;
+		
+		Primitive greenx = new Int4ByteUnsignedBE();		
+		PrimitiveData greenxData;
+		
+		Primitive greeny = new Int4ByteUnsignedBE();		
+		PrimitiveData greenyData;
+		
+		Primitive bluex = new Int4ByteUnsignedBE();		
+		PrimitiveData bluexData;
+		
+		Primitive bluey = new Int4ByteUnsignedBE();		
+		PrimitiveData blueyData;
+
+		/*
+		 * READ FROM FILE TO FILL PRIMITIVE DATA
+		 */
+		boolean keepgoing = true;
+		if (dataendpos >= (raf.getFilePointer() + 4)) {
+			try {
+				long bytepos = raf.getFilePointer();
+				whitepointx.setValueFromFile(raf);
+				whitepointxData = new PrimitiveData(whitepointx, KlangConstants.PRIMITIVE_DATA_CHRM_CHUNK_WHITEPOINTX, bytepos, raf.getFilePointer());
+			} catch (BadValueException er) {
+				whitepointxData = new PrimitiveData(whitepointx, KlangConstants.PRIMITIVE_DATA_CHRM_CHUNK_WHITEPOINTX, null, null);
+			}
+		} else {
+			whitepointxData = new PrimitiveData(whitepointx, KlangConstants.PRIMITIVE_DATA_CHRM_CHUNK_WHITEPOINTX, null, null);
+		}
+		
+		if (keepgoing && (dataendpos >= (raf.getFilePointer() + 4))) {
+			try {
+				long bytepos = raf.getFilePointer();
+				whitepointy.setValueFromFile(raf);
+				whitepointyData = new PrimitiveData(whitepointy, KlangConstants.PRIMITIVE_DATA_CHRM_CHUNK_WHITEPOINTY, bytepos, raf.getFilePointer());
+			} catch (BadValueException er) {
+				whitepointyData = new PrimitiveData(whitepointy, KlangConstants.PRIMITIVE_DATA_CHRM_CHUNK_WHITEPOINTY, null, null);
+			}
+		} else {
+			whitepointyData = new PrimitiveData(whitepointy, KlangConstants.PRIMITIVE_DATA_CHRM_CHUNK_WHITEPOINTY, null, null);
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 67:77, 79:89, 91:101, 103:113, 115:125, 127:137, 139:149.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.